homeassistant/components/amcrest/binary_sensor.py

```diff
@@ -73,17 +73,17 @@ def check_binary_sensors(value):
     return value
 
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+async def async_setup_entry(hass, config_entry, async_add_entities):
     """Set up a binary sensor for an Amcrest IP Camera."""
-    if discovery_info is None:
+    if config_entry is None:
         return
 
-    name = discovery_info[CONF_NAME]
+    name = config_entry.data[CONF_NAME]
     device = hass.data[DATA_AMCREST][DEVICES][name]
     async_add_entities(
         [
             AmcrestBinarySensor(name, device, sensor_type)
-            for sensor_type in discovery_info[CONF_BINARY_SENSORS]
+            for sensor_type in config_entry.options.get(CONF_BINARY_SENSORS)
```

Comment:
We should remove the option to define which binary sensors are added. Instead that should be controlled by the user by enabling/disabling entities in the entity registry.

Reply:
not all device / firmware versions support all binary sensors. some support some, others support others

Reply:
There are indeed way too many different specific "events" the cameras can issue to add them all.   Some are determined by what options are enabled in the camera itself and some are undocumented but essential for triggering events such as doorbell button presses.

Currently, only 2 cameras events are defined as binary sensors - audio_detected and motion_detected  (plus their *_polled alternatives).   The new capability that blademckain is adding is not to define new binary sensors, but to use the event platform to recognize any and all event codes the cameras can throw.   Here are some of the various _documented_ events a camera can issue (newer AI cameras have many more, older ones will have fewer): 
```
events[0]=LoginFailure
events[1]=VideoMotion
events[2]=VideoBlind
events[3]=VideoAbnormalDetection
events[4]=VideoUnFocus
events[5]=FaceDetection
events[6]=CrossLineDetection
events[7]=CrossRegionDetection
events[8]=LeftDetection
events[9]=TakenAwayDetection
events[10]=AudioMutation
events[11]=AudioAnomaly
events[12]=StorageNotExist
events[13]=StorageFailure
events[14]=StorageLowSpace
```

This is incredibly powerful and flexible, but very different from how sensors are usually defined.   You have to know the exact code syntax to define the new event by that very name.    

For example, to set up a sensor for motion detection, you can  use the motion_detected binary sensor, which under the covers is subscribing to the "VideoMotion" event.    Or you can use the proposed new amcrest event platform to define "amcrest.VideoMotion".    Alternatively, you could define amcrest.CrossLineDetection to subscribe to "Tripwire" events instead of or in addition to VideoMotion.      

Furthermore each of these event codes includes an "Action" - such as "Start", "Stop", "Pulse", "Invite", "Hangup".  There may also be an "Index" code.   Are these included in the attributes of each event?  It makes automation a bit trickier to handle since a condition on each action will likely be needed along with the trigger.   If you don't take the action into account, you will get a lot of unwanted event triggers when you only wanted when the event to trigger once.   

So...  if I wanted to create a doorbell press sensor for the Amcrest AD110, In the UI, I would define a config flow event field with the undocumented and esoteric "**\__DoTalkAction\__**"  and then in an automation, use something like this  (assuming the event code isn't case sensitive and there actually is a case-sensitive action attribute)?
```
# Proposed AD110 Doorbell button event 
 trigger:
    platform: event
    event_type: amcrest._dotalkaction_
  condition:
     - condition: template   
       value_template: "{{ is_state('event.amcrest._dotalkaction_.attributes.action', 'Invite') }}"
```
Is that right?   It's a bit awkward it seems to me.  I think it would be better to define a button press as an actual binary sensor and use of the event platform an optional way of doing more advanced things.

```
# AD110 Doorbell button binary sensor alternative
 trigger:
     platform: state
     entity_id: 
       - binary_sensor.amcrest_button_detected
      to: 'on'
```
There may be other binary sensors that would be sensible to define versus using solely this event platform.   There is room for both methods I believe, and it's OK if there can be 2 ways to process events.  There already will be with Motion_Detected.

Reply:
Events should not be binary sensors to begin with. All events should be … events. 

See https://developers.home-assistant.io/docs/integration_events